### PR TITLE
Add -s option, like -m, -s sets the console-script to execute

### PR DESCRIPTION
**Pull request by izissise**
Today the only way to choose console-scripts after the build is
complete is to set PEX_SCRIPT env var.

### CODE REVIEW

File: pex/pex.py [hunk 1]
```diff
@@ -585,6 +585,10 @@ def execute_interpreter(self):
                 module = args[1]
                 sys.argv = args[1:]
                 return self.execute_module(module, alter_sys=True)
+            elif arg == "-s":
```
Comment (jsirois) [inline pex/pex.py:588]: The execute_interpreter mode is intended to act just like `python` as much as possible. The `-s` option means something else for `python` and `python` doesn't even have the concept of a console script - that's a higher level thing added by `setuptools` distributions.

You may know this stuff, so forgive me if you do - but just to run down all the options today:
1. If you just want to always use 1 console script, build the pex with `-c` / `--script` / `--console-script`.
2. If you want to use an add hoc script on the CLI just use `PEX_SCRIPT=foo ./my.pex ...`.
3. As an alternative to 2, build your pex with `--include-tools` and then run `PEX_TOOLS=1 ./my.pex venv create/a/venv/from/the/my.pex/right/here` and then source the activate script and run any console script you want.

Comment (izissise) [reply pex/pex.py:588]: I have a use case for a project where there are multiples console-script defined, and I want a user to be able to choose which one their wants at runtime.
Using PEX_SCRIPT works fine, I just find that an option parameter would have a better usability in this case.

**Comment by jsirois** [comment]
@izissise you can do much better. What do you think of this?:

You'd create something like this  - here in `src/pex_busybox.py`:
```python 
import argparse
import code
import os
import sys
from collections import OrderedDict
from importlib.metadata import entry_points


def main():
    argv0 = sys.argv[0]
    exe_name = os.path.basename(argv0)
    console_scripts = OrderedDict(
        (ep.name, ep)
        for ep in sorted(entry_points().get("console_scripts", ()), key=lambda ep: ep.name)
    )

    # BusyBox-style execution (dispatch based on link name):
    if exe_name in console_scripts:
        os.environ["PEX_SCRIPT"] = exe_name
        os.execv(argv0, sys.argv)

    # Manual entrypoint execution with extra ergonomics:
    parser = argparse.ArgumentParser(description=f"A {exe_name} PEX busy box.")
    subparsers = parser.add_subparsers()
    for console_script in console_scripts:
        subparser = subparsers.add_parser(console_script)
        subparser.set_defaults(script=console_script)

    args, rest = parser.parse_known_args()
    script = vars(args).get("script")
    if script:
        os.environ["PEX_SCRIPT"] = script
        os.execv(argv0, [argv0, *rest])

    # PEX interpreter style execution (single file venv repl):
    code.interact()


if __name__ == "__main__":
    main()
```

Then build a PEX using that file as your entry point and include whatever distributions you want. Here I use `isort` and `black`:
```
$ pex -D src -e pex_busybox:main isort black -o busybox.pex
```
That gives you a PEX file with multiple, _discoverable_, entrypoints. Presumably this is better than adding `-s` into Pex itself and then having to tell users what scripts were available in your particular PEX file out of band:

PEX file venv-repl mode:
```
$ ./busybox.pex 
Python 3.9.5 (default, May 24 2021, 12:50:35) 
[GCC 11.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import isort
>>> import black
>>> 
now exiting InteractiveConsole...
```
Self discovery of entrypoints in the most obvious way:
```
$ ./busybox.pex -h
usage: busybox.pex [-h] {black,black-primer,blackd,isort,isort-identify-imports} ...

A busybox.pex PEX busy box.

positional arguments:
  {black,black-primer,blackd,isort,isort-identify-imports}

optional arguments:
  -h, --help            show this help message and exit
```
Actually use one of those scripts listed via `--help`:
```
$ ./busybox.pex black --version
python -m busybox, version 21.5b2
```
And you can setup BusyBox style aliases:
```
$ ln -s busybox.pex isort
$ ./isort --version

                 _                 _
                (_) ___  ___  _ __| |_
                | |/ _/ / _ \/ '__  _/
                | |\__ \/\_\/| |  | |_
                |_|\___/\___/\_/   \_/

      isort your imports, so you don't have to.

                    VERSION 5.8.0

```

**Comment by izissise** [comment]
Wow, awesome that exactly the usage I need! Thanks :)

Is there a chance that it could be integrated as a new build flag (--busybox?), so it's directly available at build time? I'm willing to do the necessary changes in a new MR.

**Comment by jsirois** [comment]
It could, but my sketch would need cleanup in at least a few aspects:
1. The `importlib.metadata` stdlib module is not available for Python < 3.8. It can be added for Python 3.6 & 3.7 via the `importlib-metadata` project but this still doesn't cover Python 2.7 or Python 3.5 - both of which Pex supports. For those, `pkg_resources` from the `setuptools` project would need to be used. So the script would have to grow conditional imports and code as well as conditional 3rdparty requirements.
2. The technique used to execute the console scripts is in-efficient. At least 50ms, but more likely ~100ms is wasted re-executing the PEX file instead of just running the entry-point.

These items combined point to just creating an independent `python-busybox` project since, after item 2 is addressed, the code will be completely free of Pex references.

**Comment by jsirois** [comment]
Furthering this line of thought - this is getting close to correct:

src/busybox.py:
```python
$ cat src/busybox.py 
import argparse
import code
import os
import sys
from collections import OrderedDict

try:
    try:
        from importlib.metadata import entry_points
    except ImportError:
        from importlib_metadata import entry_points

    def iter_console_scripts():
        return entry_points().get("console_scripts", ())

except ImportError:
    from pkg_resources import iter_entry_points
    
    def iter_console_scripts():
        return iter_entry_points("console_scripts")


def main():
    argv0 = sys.argv[0]
    exe_name = os.path.basename(argv0)
    console_scripts = OrderedDict(
        (ep.name, ep)
        for ep in sorted(iter_console_scripts(), key=lambda ep: ep.name)
    )

    # BusyBox-style execution (dispatch based on link name):
    ep = console_scripts.get(exe_name)
    if ep:
        return ep.load()() 

    # Manual entry point execution with extra ergonomics:
    parser = argparse.ArgumentParser(
        description="A {exe_name} busy box.".format(exe_name=exe_name)
    )
    subparsers = parser.add_subparsers()
    for console_script, ep in console_scripts.items():
        subparser = subparsers.add_parser(console_script, description=console_script)
        subparser.set_defaults(ep=ep)

    args, rest = parser.parse_known_args()
    ep = vars(args).get("ep")
    if ep:
        sys.argv = [argv0] + rest
        return ep.load()()
        
    # No entry point specified; so drop to a repl:
    code.interact()


if __name__ == "__main__":
    main()
```

And then to build a multi-interpreter busybox using Pex:
```
$ pex \
--interpreter-constraint ">=2.7" \
--python-shebang "/usr/bin/env python" \
'setuptools; python_version <= "3.5"' \
'importlib-metadata; python_version == "3.6" or python_version == "3.7"' \
isort \
coverage \
-D src/ \
-ebusybox:main \
-obusybox.pex
```

If busybox (or python-busybox) were a distribution out on PyPI; this would just be:
```
$ pex \
--interpreter-constraint ">=2.7" \
--python-shebang "/usr/bin/env python" \
busybox \
isort \
coverage \
-cbusybox \
-obusybox.pex
```
So nothing built into Pex itself, but you can turn any PEX file into a busybox by adding the `busybox` requirements and specifying an entry point console script of `-cbusybox`.

**Comment by izissise** [comment]
Wow it's awesome that it can be done without depending on pex!
Thanks a lot for this !

**Comment by jsirois** [comment]
Alright, you can now use `conscript` to do this: https://pypi.org/project/conscript/

```
$ pex \
--interpreter-constraint ">=2.7" \
--python-shebang "/usr/bin/env python" \
conscript \
isort \
coverage \
-c conscript \
-o busybox.pex
```

Which nets:
```
$ ./busybox.pex -h
usage: busybox.pex [-h] [PROGRAM]

A busybox.pex busy box.

positional arguments:
  PROGRAM     The program to execute.
              
              The following programs are available:
              + coverage
              + coverage-3.9
              + coverage3
              + isort
              + isort-identify-imports

optional arguments:
  -h, --help  Show this help message and exit.
^jsirois@gill ~/dev/jsirois/conscript (main) $ ./busybox.pex isort --version

                 _                 _
                (_) ___  ___  _ __| |_
                | |/ _/ / _ \/ '__  _/
                | |\__ \/\_\/| |  | |_
                |_|\___/\___/\_/   \_/

      isort your imports, so you don't have to.

                    VERSION 5.9.1

```

Of course this also works with shiv (and anything else):
```
$ shiv \
-p "/usr/bin/env python" \
conscript \
isort \
coverage \
-c conscript \
-o busybox.shiv
Collecting conscript
  Using cached conscript-0.1.1-py2.py3-none-any.whl (7.5 kB)
Collecting isort
  Using cached isort-5.9.1-py3-none-any.whl (105 kB)
Collecting coverage
  Using cached coverage-5.5-cp39-cp39-manylinux2010_x86_64.whl (243 kB)
Installing collected packages: isort, coverage, conscript
Successfully installed conscript-0.1.1 coverage-5.5 isort-5.9.1
```
Which nets:
```
$ ./busybox.shiv -h
usage: busybox.shiv [-h] [PROGRAM]

A busybox.shiv busy box.

positional arguments:
  PROGRAM     The program to execute.
              
              The following programs are available:
              + coverage
              + coverage-3.9
              + coverage3
              + easy_install
              + easy_install-3.8
              + isort
              + isort-identify-imports
              + pip
              + pip3
              + pip3.8
              + shiv
              + shiv-info

optional arguments:
  -h, --help  Show this help message and exit.
$ ./busybox.shiv pip --version
pip 21.1.3 from /home/jsirois/.venv/shiv/lib/python3.9/site-packages/pip (python 3.9)
```

**Pull request closed without merging**